In [1]:
import pandas as pd
import numpy as np
import datetime
#import pandas_profiling # pip install pandas_profiling

taxiCancellationScore = pd.read_csv('Kaggle_YourCabs_score.csv')
taxiCancellationTraining = pd.read_csv('Kaggle_YourCabs_training.csv')

In [2]:
#taxiCancellationScore.Profile_Report()

In [3]:
taxiCancellationTraining.columns

Index(['id', 'user_id', 'vehicle_model_id', 'package_id', 'travel_type_id',
       'from_area_id', 'to_area_id', 'from_city_id', 'to_city_id', 'from_date',
       'to_date', 'online_booking', 'mobile_site_booking', 'booking_created',
       'from_lat', 'from_long', 'to_lat', 'to_long', 'Car_Cancellation',
       'Cost_of_error'],
      dtype='object')

In [4]:
taxiCancellationTraining.isnull().sum()

id                         0
user_id                    0
vehicle_model_id           0
package_id             35881
travel_type_id             0
from_area_id              88
to_area_id              9138
from_city_id           27086
to_city_id             41843
from_date                  0
to_date                17890
online_booking             0
mobile_site_booking        0
booking_created            0
from_lat                  93
from_long                 93
to_lat                  9138
to_long                 9138
Car_Cancellation           0
Cost_of_error              0
dtype: int64

### Drop 9138 rows, Drop from_city_id, Drop row from_lat

In [5]:
taxiCancellationScore.isnull().sum()

id                         0
user_id                    0
vehicle_model_id           0
package_id              8215
travel_type_id             0
from_area_id              20
to_area_id              2150
from_city_id            6288
to_city_id              9635
from_date                  0
to_date                 4127
online_booking             0
mobile_site_booking        0
booking_created            0
from_lat                  20
from_long                 20
to_lat                  2150
to_long                 2150
Unnamed: 18            10000
Unnamed: 19            10000
dtype: int64

#Convert Date String into Timestamp()
def date_time(y,x):
    from_date= [None]*y.shape[0]
    for i in range(y.shape[0]):
        if (type(y[x][i] is str and y[x][i].isnull() == False)):
            try:
                datetime_ob = datetime.datetime.strptime(y[x][i],"%m/%d/%Y %H:%M").weekday()
            except:
                print(int)
            y.loc[:,x].replace(to_replace=y.loc[:,x].values[i], value = datetime_ob, inplace = True)

In [6]:
#convert try 2
from_weekdate_test = [] #days of week
from_time_test = []
for x in taxiCancellationScore['from_date']:
    if type(x) is str:
        ziz = x.split(' ')
        day = ziz[0].split('/')
        for z in range(len(day)):
            day[z] = int(day[z])
        from_weekdate_test.append(datetime.date(day[2], day[0], day[1]).weekday())
        #from_weekdate is now ints as day of week, 0=sunday etc

        time = ziz[1].split(':')
        from_time_test.append(int(time[0])*60 + int(time[1]))
        #from_time is now ints as minute of day
        
        
booking_weekdate_test = [] #days of week
booking_time_test = []
for x in taxiCancellationScore['booking_created']:
    if type(x) is str:
        ziz = x.split(' ')
        day = ziz[0].split('/')
        for z in range(len(day)):
            day[z] = int(day[z])
        booking_weekdate_test.append(datetime.date(day[2], day[0], day[1]).weekday())
        #booking_weekdate is now ints as day of week, 0=sunday etc

        time = ziz[1].split(':')
        booking_time_test.append(int(time[0])*60 + int(time[1]))
        #booking_time is now ints as minute of day
        
#to be appended: from_weekdate, from_time, booking_weekdate, booking_time

In [7]:
#convert try 2
from_weekdate = [] #days of week
from_time = []
for x in taxiCancellationTraining['from_date']:
    if type(x) is str:
        ziz = x.split(' ')
        day = ziz[0].split('/')
        for z in range(len(day)):
            day[z] = int(day[z])
        from_weekdate.append(datetime.date(day[2], day[0], day[1]).weekday())
        #from_weekdate is now ints as day of week, 0=sunday etc

        time = ziz[1].split(':')
        from_time.append(int(time[0])*60 + int(time[1]))
        #from_time is now ints as minute of day
        
        
booking_weekdate = [] #days of week
booking_time = []
for x in taxiCancellationTraining['booking_created']:
    if type(x) is str:
        ziz = x.split(' ')
        day = ziz[0].split('/')
        for z in range(len(day)):
            day[z] = int(day[z])
        booking_weekdate.append(datetime.date(day[2], day[0], day[1]).weekday())
        #booking_weekdate is now ints as day of week, 0=sunday etc

        time = ziz[1].split(':')
        booking_time.append(int(time[0])*60 + int(time[1]))
        #booking_time is now ints as minute of day
        
#to be appended: from_weekdate, from_time, booking_weekdate, booking_time

In [8]:
date_df = pd.DataFrame({'from_weekdate':from_weekdate,'from_time':from_time,'booking_weekdate':booking_weekdate,'booking_time':booking_time})
new_taxiCancellationTraining = taxiCancellationTraining.join(date_df)

#Replace Trained Value in DataFrame
date_time(taxiCancellationScore,'from_date')
taxiCancellationScore.loc[:,'from_date']

date_time(taxiCancellationScore,'booking_created')
taxiCancellationScore.loc[:,'booking_created']

## Addressing Null values

In [9]:
new_taxiCancellationTraining.dropna(subset=['to_area_id', 'to_lat', 'to_long', ], inplace = True)

In [10]:
new_taxiCancellationTraining.isnull().sum()

id                         0
user_id                    0
vehicle_model_id           0
package_id             34293
travel_type_id             0
from_area_id               0
to_area_id                 0
from_city_id           19788
to_city_id             34293
from_date                  0
to_date                15186
online_booking             0
mobile_site_booking        0
booking_created            0
from_lat                   0
from_long                  0
to_lat                     0
to_long                    0
Car_Cancellation           0
Cost_of_error              0
from_weekdate              0
from_time                  0
booking_weekdate           0
booking_time               0
dtype: int64

In [11]:
new_taxiCancellationTraining

id  user_id  vehicle_model_id  package_id  travel_type_id  \
0      132512    22177                28         NaN               2   
1      132513    21413                12         NaN               2   
2      132514    22178                12         NaN               2   
3      132515    13034                12         NaN               2   
4      132517    22180                12         NaN               2   
5      132518    17712                12         NaN               2   
7      132520    22181                12         NaN               2   
9      132522    22184                12         NaN               2   
10     132523     4941                12         NaN               2   
11     132524    17037                12         NaN               2   
12     132525    22185                12         NaN               2   
13     132526    22186                12         NaN               2   
14     132527       16                28         NaN               2   
15     132528    15097                12         NaN               2   
16     132530      761                12         NaN               2   
18     132533    17226                12         NaN               2   
19     132534      868                12         NaN               2   
21     132536       16                28         NaN               2   
22     132537    21558                12         NaN               2   
25     132540    22193                12         NaN               2   
26     132541    20430                12         NaN               2   
27     132542    22194                12         NaN               2   
28     132543    21716                28         NaN               2   
29     132544    38966                12         NaN               2   
30     132545    12973                12         NaN               2   
31     132546    22196                12         NaN               2   
32     132550     8327                12         NaN               2   
33     132551    22200                12         NaN               2   
35     132553    22202                12         NaN               2   
36     132555    20340                12         NaN               2   
...       ...      ...               ...         ...             ...   
43378  185881    48690                12         NaN               2   
43380  185884    48692                12         NaN               2   
43381  185885    48693                12         NaN               2   
43382  185886    48694                12         NaN               2   
43383  185887    48696                12         NaN               2   
43384  185888     4367                12         NaN               2   
43388  185893    15932                12         NaN               2   
43389  185895    48701                87         NaN               2   
43391  185897    48703                12         NaN               2   
43392  185898    39618                12         NaN               2   
43394  185900    48705                12         NaN               2   
43395  185901    48539                89         NaN               2   
43396  185902    48706                12         NaN               2   
43397  185903    48707                12         NaN               2   
43399  185907    48709                12         NaN               2   
43404  185912    48711                12         NaN               2   
43412  185921    44027                12         NaN               2   
43413  185922    48718                12         NaN               2   
43414  185923    34127                12         NaN               2   
43416  185926    36929                12         NaN               2   
43418  185928    48721                12         NaN               2   
43419  185929    46807                12         NaN               2   
43421  185931    48723                12         NaN               2   
43422  185932    48724                12         Na

In [12]:
feature_cols = ['vehicle_model_id', 'travel_type_id',
                'from_area_id', 'to_area_id', 
       'online_booking', 'mobile_site_booking',
       'from_lat', 'from_long', 'to_lat', 'to_long', 'from_weekdate','from_time','booking_weekdate','booking_time']

In [13]:
X = new_taxiCancellationTraining.loc[:,feature_cols] # Features
y = new_taxiCancellationTraining.loc[:,'Car_Cancellation'] # Target variable
X.isnull().sum()

vehicle_model_id       0
travel_type_id         0
from_area_id           0
to_area_id             0
online_booking         0
mobile_site_booking    0
from_lat               0
from_long              0
to_lat                 0
to_long                0
from_weekdate          0
from_time              0
booking_weekdate       0
booking_time           0
dtype: int64

In [14]:
X.shape
#34k rows

(34293, 14)

In [15]:
y.isnull().sum()

0

## Not append the Test dataframe yet

In [16]:
date_df_test = pd.DataFrame({'from_weekdate':from_weekdate_test,'from_time':from_time_test,'booking_weekdate':booking_weekdate_test,'booking_time':booking_time_test})
new_taxiCancellationScore =taxiCancellationScore.join(date_df_test)

In [17]:
new_taxiCancellationScore.dropna(subset=['to_area_id', 'to_lat', 'to_long', ], inplace = True)
X_test = new_taxiCancellationScore.loc[:,feature_cols] # Features
X_test.isnull().sum()

vehicle_model_id       0
travel_type_id         0
from_area_id           0
to_area_id             0
online_booking         0
mobile_site_booking    0
from_lat               0
from_long              0
to_lat                 0
to_long                0
from_weekdate          0
from_time              0
booking_weekdate       0
booking_time           0
dtype: int64

## 4 Models

In [19]:
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Dense, Dropout

#Logistic Regresstion

ModuleNotFoundError: No module named 'tensorflow'

In [ ]:
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression(random_state=0, solver='lbfgs', multi_class='multinomial')
logreg.fit(X, y)

In [ ]:
#Deep Learning Keras Model
#relu -linear
#sigmoid function for final layer
#Dropout 50%
#Dimension input features column
#Output binary classification 0 and 1
model = Sequential()
model.add(Dense(124, activation='relu', input_dim = X.shape[1]))
model.add(Dropout(0.3))
model.add(Dense(64, activation='relu'))
model.add(Dropout(0.3))
model.add(Dense(32, activation='relu'))
model.add(Dropout(0.3))
model.add(Dense(1, activation='sigmoid'))
model.summary()
#Using binary_crossentropy
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])


In [ ]:
#KNN
n_neigh = 3
from sklearn.neighbors import KNeighborsClassifier
neighbor = KNeighborsClassifier(n_neigh)
neighbor.fit(X, y) 

In [ ]:
#Naive Bayes
from sklearn.naive_bayes import GaussianNB
gaus = GaussianNB()
gaus.fit(X, y)

# Prediction

In [ ]:
model.fit(X,y, epochs=10, batch_size = 2, verbose = 1)

In [ ]:
model.save('trained_model.h5')

In [ ]:
X_test.isnull().sum()

In [ ]:
y_pred_logreg = logreg.predict(X_test)
y_pred_keras= model.predict(X_test)
y_pred_neigh= neighbor.predict(X_test)
y_pred_gaus= gaus.predict(X_test)

# Performance of Each Model

# visualization

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [ ]:
from sklearn import metrics
import numpy as np

cnf_matrixlogreg = metrics.confusion_matrix(y, y_pred_logreg)
cnf_matrixkeras = metrics.confusion_matrix(y, y_pred_keras)
cnf_matrixneigh = metrics.confusion_matrix(y, y_pred_neigh)
cnf_matrixgaus = metrics.confusion_matrix(y, y_pred_gaus)
cnf_matrixlogreg
cnf_matrixkeras
cnf_matrixneigh
cnf_matrixgaus

In [ ]:
class_names=['No cancellation','Cancellation'] # name  of classes
fig, ax = plt.subplots()
tick_marks = np.arange(len(class_names))
# create heatmap
sns.heatmap(cnf_matrixlogreg, annot=True, cmap="YlGnBu" ,fmt='d', xticklabels=class_names, yticklabels=class_names)
ax.xaxis.set_label_position("top")
plt.tight_layout()
plt.title('Confusion matrix logistic regression')
plt.ylabel('Actual label')
plt.xlabel('Predicted label')



In [ ]:
class_names=['No cancellation','Cancellation'] # name  of classes
fig, ax = plt.subplots()
tick_marks = np.arange(len(class_names))
# create heatmap
sns.heatmap(cnf_matrixkeras, annot=True, cmap="YlGnBu" ,fmt='d', xticklabels=class_names, yticklabels=class_names)
ax.xaxis.set_label_position("top")
plt.tight_layout()
plt.title('Confusion matrix keras')
plt.ylabel('Actual label')
plt.xlabel('Predicted label')

In [ ]:
class_names=['No cancellation','Cancellation'] # name  of classes
fig, ax = plt.subplots()
tick_marks = np.arange(len(class_names))
# create heatmap
sns.heatmap(cnf_matrixneigh, annot=True, cmap="YlGnBu" ,fmt='d', xticklabels=class_names, yticklabels=class_names)
ax.xaxis.set_label_position("top")
plt.tight_layout()
plt.title('Confusion matrix knn')
plt.ylabel('Actual label')
plt.xlabel('Predicted label')

In [ ]:
class_names=['No cancellation','Cancellation'] # name  of classes
fig, ax = plt.subplots()
tick_marks = np.arange(len(class_names))
# create heatmap
sns.heatmap(cnf_matrixgaus, annot=True, cmap="YlGnBu" ,fmt='d', xticklabels=class_names, yticklabels=class_names)
ax.xaxis.set_label_position("top")
plt.tight_layout()
plt.title('Confusion matrix gaussian')
plt.ylabel('Actual label')
plt.xlabel('Predicted label')